File: mergeSort/mergeSort.py

def sort(li):
    if(len(li)<= 1):
        return li
    
    mid = len(li)//2
    left = li[0:mid]
    right=li[mid:]
    
    sorted_left = sort(left)
    sorted_right = sort(right)
    # print(merge([1,2],[5,4]));
    return merge(sorted_left,sorted_right)
    

def merge(left,right):
    result = []
    i=0
    j=0
    
    while(i <len(left) and j < len(right)):
        if(right[j]> left[i] ):
            result.append(left[i])
            i+=1
        else:
            result.append(right[j])
            j+=1
   
    while(i< len(left)):
        result.append(left[i])
        i+=1
    
    while(j< len(right)):
        result.append(right[j])
        j+=1
    return result


print(sort([5,4,3,2,1]))

File: mergeSort/mergeSort.js
// Merge function 
function merge(arr1, arr2) {
  let results = [];
  let i = 0;
  let j = 0;
  while (i < arr1.length && j < arr2.length) {
    if (arr2[j] > arr1[i]) {
      results.push(arr1[i]);
      i++;
    } else {
      results.push(arr2[j])
      j++;
    }
  }

  while (i < arr1.length) {
    results.push(arr1[i])
    i++;
  }
  
  while (j < arr2.length) {
    results.push(arr2[j])
    j++;
  }
  return results;
}

// Recrusive Merge Sort
function mergeSort(arr) {
  if (arr.length <= 1) return arr;
  const mid = Math.floor(arr.length / 2);
  const left = arr.slice(0, mid);
  const right = arr.slice(mid);

  const sortedLeft = mergeSort(left);
  const sortedRight = mergeSort(right);
  return merge(sortedLeft, sortedRight);
}

console.log(mergeSort([5, 24, 7, 9, 80, 76, 73]))
